nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchHttpRecord.java

```diff
@@ -0,0 +1,559 @@
+                    id = null;
+                }
+
+                // The ID must be valid for all operations except "index". For that case,
+                // a missing ID indicates one is to be auto-generated by Elasticsearch
+                if (id == null && !indexOp.equalsIgnoreCase("index")) {
+                    throw new IdentifierNotFoundException("Index operation {} requires a valid identifier value from a flow file attribute, transferring to failure.");
+                }
+
+                final StringBuilder json = new StringBuilder();
+
+                ByteArrayOutputStream out = new ByteArrayOutputStream();
+                JsonGenerator generator = factory.createJsonGenerator(out);
+                writeRecord(record, record.getSchema(), generator);
+                generator.flush();
+                generator.close();
+                json.append(out.toString());
+
+                if (indexOp.equalsIgnoreCase("index")) {
+                    sb.append("{\"index\": { \"_index\": \"");
+                    sb.append(index);
+                    sb.append("\", \"_type\": \"");
+                    sb.append(docType);
+                    sb.append("\"");
+                    if (!StringUtils.isEmpty(id)) {
+                        sb.append(", \"_id\": \"");
+                        sb.append(id);
+                        sb.append("\"");
+                    }
+                    sb.append("}}\n");
+                    sb.append(json);
+                    sb.append("\n");
+                } else if (indexOp.equalsIgnoreCase("upsert") || indexOp.equalsIgnoreCase("update")) {
+                    sb.append("{\"update\": { \"_index\": \"");
+                    sb.append(index);
+                    sb.append("\", \"_type\": \"");
+                    sb.append(docType);
+                    sb.append("\", \"_id\": \"");
+                    sb.append(id);
+                    sb.append("\" }\n");
+                    sb.append("{\"doc\": ");
+                    sb.append(json);
+                    sb.append(", \"doc_as_upsert\": ");
+                    sb.append(indexOp.equalsIgnoreCase("upsert"));
+                    sb.append(" }\n");
+                } else if (indexOp.equalsIgnoreCase("delete")) {
+                    sb.append("{\"delete\": { \"_index\": \"");
+                    sb.append(index);
+                    sb.append("\", \"_type\": \"");
+                    sb.append(docType);
+                    sb.append("\", \"_id\": \"");
+                    sb.append(id);
+                    sb.append("\" }\n");
+                }
+            }
+        } catch (IdentifierNotFoundException infe) {
+            logger.error(infe.getMessage(), new Object[]{flowFile});
+            flowFile = session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+
+        } catch (final IOException | SchemaNotFoundException | MalformedRecordException e) {
```

Comment:
Oops, I copied that from AbstractRecordProcessor (which is used by ConvertRecord and UpdateRecord), but it throws ProcessException because it's inside a StreamCallback, and is handled "outside" by routing to failure. I will change this to route to failure, good catch!